lib/ocpp/v201/database_handler.cpp

```diff
@@ -743,18 +759,197 @@ void DatabaseHandler::insert_or_update_charging_profile(const int evse_id, const
+    return this->database->clear_table("CHARGING_PROFILES");
+}
+
+bool DatabaseHandler::clear_charging_profiles_matching_criteria(const std::optional<int32_t> profile_id,
+                                                                const std::optional<ClearChargingProfile>& criteria) {
+    // K10.FR.03, K10.FR.09
+    if (profile_id.has_value()) {
+        return this->delete_charging_profile(profile_id.value());
+    }
+
+    // criteria has no value, so clear all
+    if (!criteria.has_value()) {
+        return this->clear_charging_profiles();
+    }
+
+    if (criteria->chargingProfilePurpose.has_value() || criteria->evseId.has_value() ||
+        criteria->stackLevel.has_value()) {
+        std::string delete_query = "DELETE FROM CHARGING_PROFILES";
+        // Start with K10.FR.04, prevent deleting external constraints
+        std::vector<std::string> filters = {"CHARGING_PROFILE_PURPOSE != 'ChargingStationExternalConstraints'"};
+
+        if (criteria->chargingProfilePurpose.has_value()) {
+            filters.push_back("CHARGING_PROFILE_PURPOSE = @charging_profile_purpose");
+        }
+
+        if (criteria->stackLevel.has_value()) {
+            filters.push_back("STACK_LEVEL = @stack_level");
+        }
+
+        if (criteria->evseId.has_value()) {
+            filters.push_back("EVSE_ID = @evse_id");
+        }
+
+        delete_query += " WHERE " + boost::algorithm::join(filters, " AND ");
+
+        auto stmt = this->database->new_statement(delete_query);
+        if (criteria->chargingProfilePurpose.has_value()) {
+            stmt->bind_text(
+                "@charging_profile_purpose",
+                conversions::charging_profile_purpose_enum_to_string(criteria->chargingProfilePurpose.value()),
+                SQLiteString::Transient);
+        }
+
+        if (criteria->stackLevel.has_value()) {
+            stmt->bind_int("@stack_level", criteria->stackLevel.value());
+        }
+
+        if (criteria->evseId.has_value()) {
+            stmt->bind_int("@evse_id", criteria->evseId.value());
+        }
+
+        if (stmt->step() != SQLITE_DONE) {
+            throw QueryExecutionException(this->database->get_error_message());
+        }
+
+        return stmt->changes() > 0;
+    }
+
+    return false;
+}
+
+std::vector<std::tuple<int, v201::ChargingProfile, ChargingLimitSourceEnum>>
+DatabaseHandler::get_charging_profiles_matching_criteria(const std::optional<int32_t> evse_id,
+                                                         const ChargingProfileCriterion& criteria) {
+    auto results = std::vector<std::tuple<int, v201::ChargingProfile, ChargingLimitSourceEnum>>();
+
+    std::string select_stmt = "SELECT EVSE_ID, PROFILE, CHARGING_LIMIT_SOURCE FROM CHARGING_PROFILES";
+    std::vector<std::string> where_clauses;
+
+    if (evse_id.has_value()) {
+        where_clauses.push_back("EVSE_ID = @evse_id");
+    }
+
+    if (criteria.chargingProfileId.has_value() && !criteria.chargingProfileId->empty()) {
```

Comment:
Maybe this entire block can be extracted into a separate function, so it becomes something like:

if (criteria has specific profile IDs) then return get_specific_charging_profiles(evse_id, criteria);

Reply:
And then the remainder of the code can be refactored into a separate function.

So it would look like:
if (criteria has specific profile IDs) {
   return get_specific_charging_profiles(evse_id, criteria);
}

return get_charging_profiles_by_criteria(evse_id, criteria);


Not happy with the name of the last one but it might make this function easier to understand.

Reply:
We want to refactor this whole setup/the database handler significantly later, which will help simplify this function.